Apps/Sandcastle/gallery/Custom Per-Feature Post Process.html

```diff
@@ -0,0 +1,99 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+    <meta charset="utf-8">
+    <meta http-equiv="X-UA-Compatible" content="IE=edge">
+    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1, minimum-scale=1, user-scalable=no">
+    <meta name="description" content="Custom per-feature post processing effect.">
+    <meta name="cesium-sandcastle-labels" content="Showcases, Post Processing">
+    <title>Cesium Demo</title>
+    <script type="text/javascript" src="../Sandcastle-header.js"></script>
+    <script type="text/javascript" src="../../../ThirdParty/requirejs-2.1.20/require.js"></script>
+    <script type="text/javascript">
+        require.config({
+            baseUrl : '../../../Source',
+            waitSeconds : 60
+        });
+    </script>
+</head>
+<body class="sandcastle-loading" data-sandcastle-bucket="bucket-requirejs.html">
+<style>
+    @import url(../templates/bucket.css);
+</style>
+<div id="cesiumContainer" class="fullSize"></div>
+<div id="loadingOverlay"><h1>Loading...</h1></div>
+<div id="toolbar"></div>
+<script id="cesium_sandcastle_script">
+function startup(Cesium) {
+    'use strict';
+//Sandcastle_Begin
+var viewer = new Cesium.Viewer('cesiumContainer', {
+    shouldAnimate : true
+});
+var scene = viewer.scene;
+
+var position = Cesium.Cartesian3.fromDegrees(-123.0744619, 44.0503706);
+var url = '../../SampleData/models/CesiumMan/Cesium_Man.glb';
+var entity = viewer.trackedEntity = viewer.entities.add({
+    name : url,
+    position : position,
+    model : {
+        uri : url
+    }
+});
+
+// Shade selected model with highlight.
+var fragmentShaderSource =
+    'uniform sampler2D colorTexture;\n' +
+    'varying vec2 v_textureCoordinates;\n' +
+    'uniform vec4 highlight;\n' +
+    'void main() {\n' +
+    '    vec4 color = texture2D(colorTexture, v_textureCoordinates);\n' +
+    '    if (czm_selected()) {\n' +
+    '        vec3 highlighted = highlight.a * highlight.rgb + (1.0 - highlight.a) * color.rgb;\n' +
+    '        gl_FragColor = vec4(highlighted, 1.0);\n' +
+    '    } else { \n' +
+    '        gl_FragColor = color;\n' +
+    '    }\n' +
+ '}\n';
+var stage = scene.postProcessStages.add(new Cesium.PostProcessStage({
+    fragmentShader : fragmentShaderSource,
+    uniforms : {
+        highlight : function() {
+            return new Cesium.Color(1.0, 0.0, 0.0, 0.5);
+        }
+    }
+}));
+stage.selected = [];
+
+var remove = scene.postRender.addEventListener(function() {
+    var primitives = scene.primitives;
+    var length = primitives.length;
+
+    var i;
+    var primitive;
+    for (i = 0; i < length; ++i) {
+        primitive = primitives.get(i);
+        if (primitive.id === entity) {
+            remove();
+            break;
```

Comment:
I guess we don't have a better way to get the primitive from an entity, but this function almost seems too complicated to be part of a Sandcastle demo. Maybe the demo should do picking like the other per-feature post processing demo?

Reply:
I thought so too, but I couldn't find any examples where we get the model from an entity. I know this was asked for a few times on the forum so now we can point to this example.

Reply:
My main concern is once it's in Sandcastle it's considered the right way to do something. Do we want to endorse this as the way to get the primitive from the entity?

Reply:
I mean, I think we on-purpose made it impossible for someone to get the primitive through the Entity API.  We generally don't want end-users to interact with the primitive if they're using the Entity API because there can be unexpected side-effects.  Maybe we should find a way to enable this via Entity API directly before adding this kind of workaround in a Sandcastle example.  @mramato  do you have any input here?

Reply:
I talked offline with @mramato when I opened the per-feature post process PR. He said we might be able to add a one-to-one mapping for entity to primitive in the future. In the meantime, this was the way to do it. Until we decide what to do I changed this to use picking like the other Sandcastle example.

Reply:
I would suggest having `scene.pick` return the entity and object describing the primitive (ie the `Entity` and `PolygonGraphics`) but we can figure that out later.  This is fine for now.

Reply:
Yeah, there are a lot of dragons here because (by design) the entity data and entity visualization are decoupled and there isn't a one-to-one mapping anywhere in the system.  A primitive knows what entity it is representing in a given frame, but an entity can be represented by any number of primitives and it has no idea what those primitives are.  There's a lot of good reasons for doing it this way, but there are obvious use cases where we say "I want the Model for this Entity".  We can certainly chose to more tightly couple entities and primitives, it would just most likely involve removing flexibility of the Entity API and also probably some breaking changes.

The possibly more correct (yet difficult) thing to do would be to better expose post processing features through the Entity API.

Reply:
For example, what about being able to associate a post-process stage to a particular ModelGraphics and everything "just works" under the hood?  I haven't dug into the post process architecture yet, so maybe that idea makes no sense.